### PR TITLE
[TT-4866] Add oas.URLRewrite data model, Fill, ExtractTo, update apidef

### PR DESCRIPTION
**Pull request by titpetric**
This PR adds the OAS definitions for URLRewrite middleware.

Observed issues:

- Enums are not supported by the current doc gen (oas.URLRewriteCondition, oas.URLRewriteInput)

Subtask: https://tyktech.atlassian.net/browse/TT-10960

### CODE REVIEW

**Comment by buger** [comment]
**API tests result - mongo44-sha256 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5933/merge
Commit: ba42926023dc1b8f09e255dc0b83353815d2438e 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7461963420)

**Comment by buger** [comment]
**API tests result - mongo44-murmur64 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5933/merge
Commit: ba42926023dc1b8f09e255dc0b83353815d2438e 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7461963420)

**Comment by buger** [comment]
**API tests result - postgres15-sha256 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5933/merge
Commit: ba42926023dc1b8f09e255dc0b83353815d2438e 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7461963420)

**Comment by buger** [comment]
**API tests result - postgres15-murmur64 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5933/merge
Commit: ba42926023dc1b8f09e255dc0b83353815d2438e 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7461963420)

File: apidef/oas/url_rewrite.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+	"fmt"
+	"sort"
+
+	"github.com/TykTechnologies/tyk/apidef"
+)
+
+// URLRewrite configures URL rewriting.
+type URLRewrite struct {
+	// Enabled enables URL rewriting if set to true.
+	Enabled bool `bson:"enabled" json:"enabled"`
```
[the remark this answers is not in the document]
Comment (titpetric) [reply apidef/oas/url_rewrite.go:13]: There is a Disabled field added there, or did you mean something else?

[the remark this answers is not in the document]
Comment (titpetric) [reply apidef/oas/url_rewrite.go:14]: Linked the main struct. Not sure we should do fine grained field-by-field, but rather only obj-obj.

File: apidef/oas/schema/x-tyk-gateway.md [hunk 2]
```diff
@@ -965,6 +968,54 @@ Path file path for the template.
...
+
+**Field: `triggers` (`[]`[URLRewriteTrigger](#urlrewritetrigger))**
+Triggers contain advanced additional triggers for the URL rewrite.
+The triggers are processed only if the requested URL matches the pattern above.
+
+
+### **URLRewriteTrigger**
+
+**Field: `condition` ([](#))**
+Condition represents a boolean operator for rules.
```
Comment (andyo-tyk) [inline apidef/oas/schema/x-tyk-gateway.md:990]:
```suggestion
Condition indicates the logical combination that will be applied to the rules for an advanced trigger:
```

File: apidef/oas/schema/x-tyk-gateway.md [hunk 3]
```diff
@@ -965,6 +968,54 @@ Path file path for the template.
...
+
+- Value `any` means any of the defined trigger rules may match
+- Value `all` means all the defined trigger rules must match
+
+**Field: `rules` (`[]`[URLRewriteRule](#urlrewriterule))**
+Rules contain conditional triggers for URL rewriting.
```
Comment (andyo-tyk) [inline apidef/oas/schema/x-tyk-gateway.md:996]:
```suggestion
Rules contain individual checks that are combined according to the `condition` to determine whether the URL rewrite will be triggered.
```

File: apidef/oas/schema/x-tyk-gateway.md [hunk 4]
```diff
@@ -965,6 +968,54 @@ Path file path for the template.
...
+If empty, it enables non-conditional rewrites.
```
Comment (andyo-tyk) [inline apidef/oas/schema/x-tyk-gateway.md:997]: Does this mean that you can have an advanced trigger without any rules?

File: apidef/oas/schema/x-tyk-gateway.md [hunk 5]
```diff
@@ -913,6 +913,9 @@ TransformRequestHeaders allows you to transform request headers.
 **Field: `transformResponseHeaders` ([TransformHeaders](#transformheaders))**
 TransformResponseHeaders allows you to transform response headers.
 
+**Field: `urlRewrite` ([URLRewrite](#urlrewrite))**
+URLRewrite contains the url rewriting configuration.
```
Comment (andyo-tyk) [inline apidef/oas/schema/x-tyk-gateway.md:917]:
```suggestion
URLRewrite contains the URL rewriting configuration.
```

**Review by andyo-tyk — DISMISSED** [review]
Suggested clarifications - pending final confirmation of the exact behaviour of the middleware

**Review by jeffy-mathew — APPROVED** [review]

File: apidef/oas/schema/x-tyk-gateway.md [hunk 6]
```diff
@@ -965,6 +968,65 @@ Path file path for the template.
...
+- `url`, match pattern against URL
+- `query`, match pattern against named query parameter value
+- `path`, match pattern against named path parameter value
+- `header`, match pattern against named header value
+- `sessionMetadata`, match pattern against session metadata
+- `requestBody`, match pattern against request body
+- `requestContext`, match pattern against request context
+
+**Field: `name` (`string`)**
+Name is the index in the input identified in `in` that should be used to locate the value for this rule. `Name` is ignored for `InputRequestBody`rules as it contains only a single value, while the others are objects.
```
Comment (andyo-tyk) [inline apidef/oas/schema/x-tyk-gateway.md:1020]: If I've understood correctly, this is actually only used for the following values in the `in` field: `query`, `path`, `header` (i.e. where in the docs above we've said "match pattern against named ... parameter value").

So perhaps (and Github isn't giving me the option to "suggest"):

Name is the index in the input identified in `in` that should be used to locate the value for this rule. `Name` is only relevant for `query`, `path` and `header` values in the `in` field, for all other values it is ignored.

**Pull request merged**